reference/config.md
# Laptop-side configuration

## Global Configuration
Telepresence uses a `config.yml` file to store and change certain global configuration values that will be used for all clusters you use Telepresence with.  The location of this file varies based on your OS:

* macOS: `$HOME/Library/Application Support/telepresence/config.yml`
* Linux: `$XDG_CONFIG_HOME/telepresence/config.yml` or, if that variable is not set, `$HOME/.config/telepresence/config.yml`

For Linux, the above paths are for a user-level configuration. For system-level configuration, use the file at `$XDG_CONFIG_DIRS/telepresence/config.yml` or, if that variable is empty, `/etc/xdg/telepresence/config.yml`.  If a file exists at both the user-level and system-level paths, the user-level path file will take precedence.

### Values

The config file currently supports values for the `timeouts`, `logLevels`, `images`, `cloud`, and `grpc` keys.

Here is an example configuration to show you the conventions of how Telepresence is configured:
**note: This config shouldn't be used verbatim, since the registry `privateRepo` used doesn't exist**

```yaml
timeouts:
  agentInstall: 1m
  intercept: 10s
logLevels:
  userDaemon: debug
images:
  registry: privateRepo # This overrides the default docker.io/datawire repo
  agentImage: ambassador-telepresence-agent:1.8.0 # This overrides the agent image to inject when intercepting
grpc:
  maxReceiveSize: 10Mi
```

#### Timeouts
Values for `timeouts` are all durations either as a number respresenting seconds or a string with a unit suffix of `ms`, `s`, `m`, or `h`.  Strings can be fractional (`1.5h`) or combined (`2h45m`).

These are the valid fields for the `timeouts` key:

|Field|Description|Default|
|---|---|---|
|`agentInstall`|Waiting for Traffic Agent to be installed|2 minutes|
|`apply`|Waiting for a Kubernetes manifest to be applied|1 minute|
|`clusterConnect`|Waiting for cluster to be connected|20 seconds|
|`intercept`|Waiting for an intercept to become active|5 seconds|
|`proxyDial`|Waiting for an outbound connection to be established|5 seconds|
|`trafficManagerConnect`|Waiting for the Traffic Manager API to connect for port fowards|20 seconds|
|`trafficManagerAPI`|Waiting for connection to the gPRC API after `trafficManagerConnect` is successful|15 seconds|

#### Log Levels
Values for `logLevels` are one of the following strings: `trace`, `debug`, `info`, `warning`, `error`, `fatal` and `panic`.
These are the valid fields for the `logLevels` key:

|Field|Description|Default|
|---|---|---|
|`userDaemon`|Logging level to be used by the User Daemon (logs to connector.log)|debug|
|`rootDaemon`|Logging level to be used for the Root Daemon (logs to daemon.log)|info|

#### Images
Values for `images` are strings. These values affect the objects that are deployed in the cluster,
so it's important to ensure users have the same configuration.

Additionally, you can deploy the server-side components with [Helm](../../install/helm), to prevent them
from being overridden by a client's config and use the [mutating-webhook](../cluster-config/#mutating-webhook)
to handle installation of the `traffic-agents`.

These are the valid fields for the `images` key:

|Field|Description|Default|
|---|---|---|
|`registry`|Docker registry to be used for installing the Traffic Manager and default Traffic Agent. If not using a helm chart to deploy server-side objects, changing this value will create a new traffic-manager deployment when using Telepresence commands. Additionally, changing this value will update installed default `traffic-agents` to use the new registry when creating a new intercept.|docker.io/datawire|
|`agentImage`|$registry/$imageName:$imageTag to use when installing the Traffic Agent. Changing this value will update pre-existing `traffic-agents` to use this new image. * the `registry` value is not used for the `traffic-agent` if you have this value set *||
|`webhookRegistry`|The container $registry that the [Traffic Manager](../cluster-config/#mutating-webhook) will use with the `webhookAgentImage` *This value is only used if a new traffic-manager is deployed*||
|`webhookAgentImage`|The container image that the [Traffic Manager](../cluster-config/#mutating-webhook) will use when installing the Traffic Agent in annotated pods *This value is only used if a new traffic-manager is deployed*||

#### Cloud
These fields control how the client interacts with the Cloud service.
Currently there is only one key and it accepts bools: `1`, `t`, `T`, `TRUE`, `true`, `True`, `0`, `f`, `F,` `FALSE`

|Field|Description|Default|
|---|---|---|
|`skipLogin`|Whether the cli should skip automatic login to Ambassador Cloud. If set to true, you must have a [license](../cluster-config/#air-gapped-cluster) installed in the cluster in order to be able to perform selective intercepts |false|

Telepresence attempts to auto-detect if the cluster is air-gapped,
be sure to set the `skipLogin` value to `true`

Reminder: To use selective intercepts, which normally require a login, you
must have a license in your cluster and specify which agentImage should be installed,
by also adding the following to your config.yml:
  ```
  images:
    agentImage: <privateRegistry>/<agentImage>
  ```

#### Grpc
The `maxReceiveSize` determines how large a message that the workstation receives via gRPC can be. The default is 4Mi (determined by gRPC). All traffic to and from the cluster is tunneled via gRPC.

The size is measured in bytes. You can express it as a plain integer or as a fixed-point number using E, G, M, or K. You can also use the power-of-two equivalents: Gi, Mi, Ki. For example, the following represent roughly the same value:
```
128974848, 129e6, 129M, 123Mi
```

## Per-Cluster Configuration
Some configuration is not global to Telepresence and is actually specific to a cluster.  Thus, we store that config information in your kubeconfig file, so that it is easier to maintain per-cluster configuration.

### Values
The current per-cluster configuration supports `dns`, `alsoProxy`, and `manager` keys.
To add configuration, simply add a `telepresence.io` entry to the cluster in your kubeconfig like so:

```
apiVersion: v1
clusters:
- cluster:
    server: https://127.0.0.1
    extensions:
    - name: telepresence.io
      extension:
        dns:
        also-proxy:
        manager:
  name: example-cluster
```
#### DNS
The fields for `dns` are: local-ip, remote-ip, exclude-suffixes, include-suffixes, and lookup-timeout.

|Field|Description|Type|Default|
|---|---|---|---|
|`local-ip`|The address of the local DNS server. This entry is only used on Linux system that are not configured to use systemd.resolved|ip|first line of /etc/resolv.conf|
|`remote-ip`|the address of the cluster's DNS service|ip|IP of the kube-dns.kube-system or the dns-default.openshift-dns service|
|`exclude-suffixes`|suffixes for which the DNS resolver will always fail (or fallback in case of the overriding resolver)|list||
|`include-suffixes`|suffixes for which the DNS resolver will always attempt to do a lookup. Includes have higher priority than excludes.|list||
|`lookup-timeout`|maximum time to wait for a cluster side host lookup|duration||

Here is an example kubeconfig:
```
apiVersion: v1
clusters:
- cluster:
    server: https://127.0.0.1
    extensions:
    - name: telepresence.io
      extension:
        dns:
          include-suffixes:
          - .se
          exclude-suffixes:
          - .com
  name: example-cluster
```


#### AlsoProxy
When using `also-proxy`, you provide a list of subnets after the key in your kubeconfig file to be added to the TUN device. All connections to addresses that the subnet spans will be dispatched to the cluster

Here is an example kubeconfig for the subnet `1.2.3.4/32`:
```
apiVersion: v1
clusters:
- cluster:
    server: https://127.0.0.1
    extensions:
    - name: telepresence.io
      extension:
        also-proxy:
        - 1.2.3.4/32
  name: example-cluster
```

#### Manager

The `manager` key contains configuration for finding the `traffic-manager` that telepresence will connect to. It supports one key, `namespace`, indicating the namespace where the traffic manager is to be found

Here is an example kubeconfig that will instruct telepresence to connect to a manager in namespace `staging`:

```yaml
apiVersion: v1
clusters:
- cluster:
    server: https://127.0.0.1
    extensions:
    - name: telepresence.io
      extension:
        manager:
          namespace: staging
  name: example-cluster
```


CHANGELOG.md
# Changelog

### 2.4.4 (TBD)

- Feature: The strategy used by traffic-manager's discovery of pod CIDRs can now be configured using the Helm chart.

- Bugfix: The traffic-manager now uses less CPU-cycles when computing the pod CIDRs.

- Feature: The agent injector now supports injecting Traffic Agents into pods that have unnamed ports.

- Bugfix: If a deployment annotated with webhook annotations is deployed before telepresence is installed, telepresence will now install an agent in that deployment before intercept

### 2.4.3 (September 15, 2021)

- Feature: The environment variable `TELEPRESENCE_INTERCEPT_ID` is now available in the interceptor's environment.

- Bugfix: A timing related bug was fixed that sometimes caused a "daemon did not start" failure.

- Bugfix: On Windows, crash stack traces and other errors were not
  written to the log files, now they are.

- Bugfix: On Linux kernel 4.11 and above, the log file rotation now
  properly reads the birth-time of the log file.  On older kernels, it
  continues to use the old behavior of using the change-time in place
  of the birth-time.

- Bugfix: Telepresence will no longer refer the user to the daemon logs for errors that aren't related to
  problems that are logged there.

- Bugfix: The overriding DNS resolver will no longer apply search paths when resolving "localhost".

- Bugfix: The cluster domain used by the DNS resolver is retrieved from the traffic-manager instead of being
  hard-coded to "cluster.local".

- Bugfix: "Telepresence uninstall --everything" now also uninstalls agents installed via mutating webhook

- Bugfix: Downloading large files during an intercept will no longer cause timeouts and hanging traffic-agent.

- Bugfix: Passing false to the intercept command's --mount flag will no longer result in a filesystem being mounted.

- Bugfix: The traffic manager will establish outbound connections in parallel instead of sequentially.

- Bugfix: The `telepresence status` command reports correct DNS settings instead of "Local IP: nil, Remote IP: nil"

### 2.4.2 (September 1, 2021)

- Feature: A new `telepresence loglevel <level>` subcommand was added that enables changing the loglevel
  temporarily for the local daemons, the `traffic-manager` and the `traffic-agents`.

- Change: The default log-level is now `info` for all components of Telepresence.

- Bugfix: The overriding DNS resolver will no longer apply search paths when resolving "localhost".

- Bugfix: The RBAC was not updated in the helm chart to enable the traffic-manager to `get` and `list`
  namespaces, which would impact users who use licensed features of the Telepresence extensions in an
  air-gapped environment.

- Bugfix: The timeout for Helm actions wasn't always respected which could cause a failing install of the
  `traffic-manager` to make the user daemon to hang indefinitely.

### 2.4.1 (August 30, 2021)

- Bugfix: Telepresence will now mount all directories from `/var/run/secrets`, not just the kubernetes.io ones.
  This allows the mounting of secrets directories such as eks.amazonaws.com (for IRSA tokens)

- Bugfix: The grpc.maxReceiveSize setting is now correctly propagated to all grpc servers.
  This allows users to mitigate a root daemon crash when sending a message over the default maximum size.

- Bugfix: Some slight fixes to the `homebrew-package.sh` script which will enable us to run
  it manually if we ever need to make homebrew point at an older version.

- Feature: Helm chart has now feature to on demand regenerate certificate used for mutating webhook by setting value.
  `agentInjector.certificate.regenerate`

- Change: The traffic-manager now requires `get` namespace permissions to get the cluster ID instead of that value being
  passed in as an environment variable to the traffic-manager's deployment.

- Change: The traffic-manager is now installed via an embedded version of the Helm chart when `telepresence connect` is first performed on a cluster.
  This change is transparent to the user.
  A new configuration flag, `timeouts.helm` sets the timeouts for all helm operations performed by the Telepresence binary.

- Bugfix: Telepresence will initialize the default namespace from the kubeconfig on each call instead of just doing it when connecting.

- Bugfix: The timeout to keep idle outbound TCP connections alive was increased from 60 to 7200 seconds which is the same as
  the Linux `tcp_keepalive_time` default.

- Bugfix: Telepresence will now remove a socket that is the result of an ungraceful termination and retry instead of printing
  an error saying "this usually means that the process has terminated ungracefully"

- Change: Failure to report metrics is logged using loglevel info rather than error.

- Bugfix: A potential deadlock situation is fixed that sometimes caused the user daemon to hang when the user
  was logged in.

- Feature: The scout reports will now include additional metadata coming from environment variables starting with
  `TELEPRESENCE_REPORT_`.

- Bugfix: The config setting `images.agentImage` is no longer required to contain the repository. The repository is
  instead picked from `images.repository`.

- Change: The `registry`, `webhookRegistry`, `agentImage` and `webhookAgentImage` settings in the `images` group of the `config.yml`
  now get their defaults from `TELEPRESENCE_AGENT_IMAGE` and `TELEPRESENCE_REGISTRY`.

### 2.4.0 (August 4, 2021)

- Feature: There is now a native Windows client for Telepresence.
  All the same features supported by the MacOS and Linux client are available on Windows.

- Feature: Telepresence can now receive messages from the cloud and raise
  them to the user when they perform certain commands.

- Bugfix: Initialization of `systemd-resolved` based DNS sets
  routing domain to improve stability in non-standard configurations.

- Bugfix: Edge case error when targeting a container by port number.
  Before if your matching/target container was at containers list index 0,

pkg/client/config.go
package client

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"github.com/sirupsen/logrus"
	"gopkg.in/yaml.v3"
	"k8s.io/apimachinery/pkg/api/resource"

	"github.com/datawire/dlib/dlog"
	"github.com/telepresenceio/telepresence/v2/pkg/client/errcat"
	"github.com/telepresenceio/telepresence/v2/pkg/filelocation"
)

const configFile = "config.yml"

// Config contains all configuration values for the telepresence CLI
type Config struct {
	Timeouts  Timeouts  `json:"timeouts,omitempty"`
	LogLevels LogLevels `json:"logLevels,omitempty"`
	Images    Images    `json:"images,omitempty"`
	Cloud     Cloud     `json:"cloud,omitempty"`
	Grpc      Grpc      `json:"grpc,omitempty"`
}

// merge merges this instance with the non-zero values of the given argument. The argument values take priority.
func (c *Config) merge(o *Config) {
	c.Timeouts.merge(&o.Timeouts)
	c.LogLevels.merge(&o.LogLevels)
	c.Images.merge(&o.Images)
	c.Cloud.merge(&o.Cloud)
	c.Grpc.merge(&o.Grpc)
}

func stringKey(n *yaml.Node) (string, error) {
	var s string
	if err := n.Decode(&s); err != nil {
		return "", errors.New(withLoc("key must be a string", n))
	}
	return s, nil
}

func (c *Config) UnmarshalYAML(node *yaml.Node) (err error) {
	if node.Kind != yaml.MappingNode {
		return errors.New(withLoc("config must be an object", node))
	}
	ms := node.Content
	top := len(ms)
	for i := 0; i < top; i += 2 {
		kv, err := stringKey(ms[i])
		if err != nil {
			return err
		}
		switch {
		case kv == "timeouts":
			err := ms[i+1].Decode(&c.Timeouts)
			if err != nil {
				return err
			}
		case kv == "logLevels":
			err := ms[i+1].Decode(&c.LogLevels)
			if err != nil {
				return err
			}
		case kv == "images":
			err := ms[i+1].Decode(&c.Images)
			if err != nil {
				return err
			}
		case kv == "cloud":
			err := ms[i+1].Decode(&c.Cloud)
			if err != nil {
				return err
			}
		case kv == "grpc":
			err := ms[i+1].Decode(&c.Grpc)
			if err != nil {
				return err
			}
		case parseContext != nil:
			dlog.Warn(parseContext, withLoc(fmt.Sprintf("unknown key %q", kv), ms[i]))
		}
	}
	return nil
}

type Timeouts struct {
	// These all nave names starting with "Private" because we "want" them to be unexported in
	// order to force you to use .TimeoutContext(), but (1) we dont' want them to be hidden from
	// the JSON/YAML engines, and (2) in the rare case, we do want to be able to reach in and
	// grab it, but we want it to be clear that this is "bad".  We should probably (TODO) get
	// rid of those later cases, but let's not spend time doing that right now; and instead just
	// make them easy to grep for (`grep Private`) later.

	// PrivateAgentInstall is how long to wait for an agent to be installed (i.e. apply of service and deploy manifests)
	PrivateAgentInstall time.Duration `json:"agentInstall,omitempty"`
	// PrivateApply is how long to wait for a k8s manifest to be applied
	PrivateApply time.Duration `json:"apply,omitempty"`
	// PrivateClusterConnect is the maximum time to wait for a connection to the cluster to be established
	PrivateClusterConnect time.Duration `json:"clusterConnect,omitempty"`
	// PrivateIntercept is the time to wait for an intercept after the agents has been installed
	PrivateIntercept time.Duration `json:"intercept,omitempty"`
	// PrivateProxyDial is how long to wait for the proxy to establish an outbound connection
	PrivateProxyDial time.Duration `json:"proxyDial,omitempty"`
	// PrivateTrafficManagerConnect is how long to wait for the traffic-manager API to connect
	PrivateTrafficManagerAPI time.Duration `json:"trafficManagerAPI,omitempty"`
	// PrivateTrafficManagerConnect is how long to wait for the initial port-forwards to the traffic-manager
	PrivateTrafficManagerConnect time.Duration `json:"trafficManagerConnect,omitempty"`
	// PrivateHelm is how long to wait for any helm operation.
	PrivateHelm time.Duration `json:"helm,omitempty"`
}

type TimeoutID int

const (
	TimeoutAgentInstall TimeoutID = iota
	TimeoutApply
	TimeoutClusterConnect
	TimeoutIntercept
	TimeoutProxyDial
	TimeoutTrafficManagerAPI
	TimeoutTrafficManagerConnect
	TimeoutHelm
)

type timeoutContext struct {
	context.Context
	timeoutID  TimeoutID
	timeoutVal time.Duration
}

func (ctx timeoutContext) Err() error {
	err := ctx.Context.Err()
	if errors.Is(err, context.DeadlineExceeded) {
		err = timeoutErr{
			timeoutID:  ctx.timeoutID,
			timeoutVal: ctx.timeoutVal,
			configFile: GetConfigFile(ctx),
			err:        err,
		}
	}
	return err
}

func (t *Timeouts) Get(timeoutID TimeoutID) time.Duration {
	var timeoutVal time.Duration
	switch timeoutID {
	case TimeoutAgentInstall:
		timeoutVal = t.PrivateAgentInstall
	case TimeoutApply:
		timeoutVal = t.PrivateApply
	case TimeoutClusterConnect:
		timeoutVal = t.PrivateClusterConnect
	case TimeoutIntercept:
		timeoutVal = t.PrivateIntercept
	case TimeoutProxyDial:
		timeoutVal = t.PrivateProxyDial
	case TimeoutTrafficManagerAPI:
		timeoutVal = t.PrivateTrafficManagerAPI
	case TimeoutTrafficManagerConnect:
		timeoutVal = t.PrivateTrafficManagerConnect
	case TimeoutHelm:
		timeoutVal = t.PrivateHelm
	default:
		panic("should not happen")
	}
	return timeoutVal
}

func (t *Timeouts) TimeoutContext(ctx context.Context, timeoutID TimeoutID) (context.Context, context.CancelFunc) {
	timeoutVal := t.Get(timeoutID)
	ctx, cancel := context.WithTimeout(ctx, timeoutVal)
	ctx = timeoutContext{
		Context:    ctx,
		timeoutID:  timeoutID,
		timeoutVal: timeoutVal,
	}
	return ctx, cancel
}

type timeoutErr struct {
	timeoutID  TimeoutID
	timeoutVal time.Duration
	configFile string
	err        error
}

func (e timeoutErr) Error() string {
	var yamlName, humanName string
	switch e.timeoutID {
	case TimeoutAgentInstall:
		yamlName = "agentInstall"
		humanName = "agent install"
	case TimeoutApply:
		yamlName = "apply"
		humanName = "apply"
	case TimeoutClusterConnect:
		yamlName = "clusterConnect"
		humanName = "cluster connect"
	case TimeoutIntercept:
		yamlName = "intercept"
		humanName = "intercept"
	case TimeoutProxyDial:
		yamlName = "proxyDial"
		humanName = "proxy dial"
	case TimeoutTrafficManagerAPI:
		yamlName = "trafficManagerAPI"
		humanName = "traffic manager gRPC API"
	case TimeoutTrafficManagerConnect:
		yamlName = "trafficManagerConnect"
		humanName = "port-forward connection to the traffic manager"
	case TimeoutHelm:
		yamlName = "helm"
		humanName = "helm operation"
	default:
		panic("should not happen")
	}
	return fmt.Sprintf("the %s timed out.  The current timeout %s can be configured as %q in %q",
		humanName, e.timeoutVal, "timeouts."+yamlName, e.configFile)
}

func (e timeoutErr) Unwrap() error {
	return e.err
}

func CheckTimeout(ctx context.Context, err error) error {
	if ctxErr := ctx.Err(); ctxErr != nil && (errors.Is(ctxErr, context.DeadlineExceeded) || err == nil) {
		return ctxErr
	}
	return err
}

// UnmarshalYAML caters for the unfortunate fact that time.Duration doesn't do YAML or JSON at all.
func (t *Timeouts) UnmarshalYAML(node *yaml.Node) (err error) {
	if node.Kind != yaml.MappingNode {
		return errors.New(withLoc("timeouts must be an object", node))
	}
	ms := node.Content
	top := len(ms)
	for i := 0; i < top; i += 2 {
		kv, err := stringKey(ms[i])
		if err != nil {
			return err
		}
		var dp *time.Duration
		switch kv {
		case "agentInstall":
			dp = &t.PrivateAgentInstall
		case "apply":
			dp = &t.PrivateApply
		case "clusterConnect":
			dp = &t.PrivateClusterConnect
		case "intercept":
			dp = &t.PrivateIntercept
		case "proxyDial":
			dp = &t.PrivateProxyDial
		case "trafficManagerAPI":
			dp = &t.PrivateTrafficManagerAPI
		case "trafficManagerConnect":
			dp = &t.PrivateTrafficManagerConnect
		case "helm":
			dp = &t.PrivateHelm
		default:
			if parseContext != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("unknown key %q", kv), ms[i]))
			}
			continue
		}

		v := ms[i+1]
		var vv interface{}
		if err = v.Decode(&vv); err != nil {
			return errors.New(withLoc("unable to parse value", v))
		}
		switch vv := vv.(type) {
		case int:
			*dp = time.Duration(vv) * time.Second
		case float64:
			*dp = time.Duration(vv * float64(time.Second))
		case string:
			if *dp, err = time.ParseDuration(vv); err != nil {
				return errors.New(withLoc(fmt.Sprintf("%q is not a valid duration", vv), v))
			}
		}
	}
	return nil
}

// merge merges this instance with the non-zero values of the given argument. The argument values take priority.
func (t *Timeouts) merge(o *Timeouts) {
	if o.PrivateAgentInstall != 0 {
		t.PrivateAgentInstall = o.PrivateAgentInstall
	}
	if o.PrivateApply != 0 {
		t.PrivateApply = o.PrivateApply
	}
	if o.PrivateClusterConnect != 0 {
		t.PrivateClusterConnect = o.PrivateClusterConnect
	}
	if o.PrivateIntercept != 0 {
		t.PrivateIntercept = o.PrivateIntercept
	}
	if o.PrivateProxyDial != 0 {
		t.PrivateProxyDial = o.PrivateProxyDial
	}
	if o.PrivateTrafficManagerAPI != 0 {
		t.PrivateTrafficManagerAPI = o.PrivateTrafficManagerAPI
	}
	if o.PrivateTrafficManagerConnect != 0 {
		t.PrivateTrafficManagerConnect = o.PrivateTrafficManagerConnect
	}
	if o.PrivateHelm != 0 {
		t.PrivateHelm = o.PrivateHelm
	}
}

type LogLevels struct {
	UserDaemon logrus.Level `json:"userDaemon,omitempty"`
	RootDaemon logrus.Level `json:"rootDaemon,omitempty"`
}

// UnmarshalYAML parses the logrus log-levels
func (ll *LogLevels) UnmarshalYAML(node *yaml.Node) (err error) {
	if node.Kind != yaml.MappingNode {
		return errors.New(withLoc("timeouts must be an object", node))
	}

	ms := node.Content
	top := len(ms)
	for i := 0; i < top; i += 2 {
		kv, err := stringKey(ms[i])
		if err != nil {
			return err
		}
		v := ms[i+1]
		level, err := logrus.ParseLevel(v.Value)
		if err != nil {
			return errors.New(withLoc("invalid log-level", v))
		}
		switch kv {
		case "userDaemon":
			ll.UserDaemon = level
		case "rootDaemon":
			ll.RootDaemon = level
		default:
			if parseContext != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("unknown key %q", kv), ms[i]))
			}
		}
	}
	return nil
}

func (ll *LogLevels) merge(o *LogLevels) {
	if o.UserDaemon != 0 {
		ll.UserDaemon = o.UserDaemon
	}
	if o.RootDaemon != 0 {
		ll.RootDaemon = o.RootDaemon
	}
}

type Images struct {
	Registry          string `json:"registry,omitempty"`
	AgentImage        string `json:"agentImage,omitempty"`
	WebhookRegistry   string `json:"webhookRegistry,omitempty"`
	WebhookAgentImage string `json:"webhookAgentImage,omitempty"`
}

// UnmarshalYAML parses the images YAML
func (img *Images) UnmarshalYAML(node *yaml.Node) (err error) {
	if node.Kind != yaml.MappingNode {
		return errors.New(withLoc("images must be an object", node))
	}

	ms := node.Content
	top := len(ms)
	for i := 0; i < top; i += 2 {
		kv, err := stringKey(ms[i])
		if err != nil {
			return err
		}
		v := ms[i+1]
		switch kv {
		case "registry":
			img.Registry = v.Value
		case "agentImage":
			img.AgentImage = v.Value
		case "webhookRegistry":
			img.WebhookRegistry = v.Value
		case "webhookAgentImage":
			img.WebhookAgentImage = v.Value
		default:
			if parseContext != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("unknown key %q", kv), ms[i]))
			}
		}
	}
	return nil
}

func (i *Images) merge(o *Images) {
	if o.AgentImage != "" {
		i.AgentImage = o.AgentImage
	}
	if o.WebhookAgentImage != "" {
		i.WebhookAgentImage = o.WebhookAgentImage
	}
	if o.Registry != "" {
		i.Registry = o.Registry
	}
	if o.WebhookRegistry != "" {
		i.WebhookRegistry = o.WebhookRegistry
	}
}

type Cloud struct {
	SkipLogin       bool          `json:"skipLogin,omitempty"`
	RefreshMessages time.Duration `json:"refreshMessages,omitempty"`
	SystemaHost     string        `json:"systemaHost,omitempty"`
	SystemaPort     string        `json:"systemaPort,omitempty"`
}

// UnmarshalYAML parses the images YAML
func (cloud *Cloud) UnmarshalYAML(node *yaml.Node) (err error) {
	if node.Kind != yaml.MappingNode {
		return errors.New(withLoc("cloud must be an object", node))
	}

	ms := node.Content
	top := len(ms)
	for i := 0; i < top; i += 2 {
		kv, err := stringKey(ms[i])
		if err != nil {
			return err
		}
		v := ms[i+1]
		switch kv {
		case "skipLogin":
			val, err := strconv.ParseBool(v.Value)
			if err != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("bool expected for key %q", kv), ms[i]))
			} else {
				cloud.SkipLogin = val
			}
		case "refreshMessages":
			duration, err := time.ParseDuration(v.Value)
			if err != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("duration expected for key %q", kv), ms[i]))
			} else {
				cloud.RefreshMessages = duration
			}
		case "systemaHost":
			cloud.SystemaHost = v.Value
		case "systemaPort":
			cloud.SystemaPort = v.Value
		default:
			if parseContext != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("unknown key %q", kv), ms[i]))
			}
		}
	}
	return nil
}

func (i *Cloud) merge(o *Cloud) {
	if o.SkipLogin {
		i.SkipLogin = o.SkipLogin
	}
	if o.RefreshMessages != 0 {
		i.RefreshMessages = o.RefreshMessages
	}
	if o.SystemaHost != "" {
		i.SystemaHost = o.SystemaHost
	}
	if o.SystemaPort != "" {
		i.SystemaPort = o.SystemaPort
	}
}

type Grpc struct {
	// MaxReceiveSize is the maximum message size in bytes the client can receive in a gRPC call or stream message.
	// Overrides the gRPC default of 4MB.
	MaxReceiveSize *resource.Quantity `json:"maxReceiveSize,omitempty"`
}

func (g *Grpc) merge(o *Grpc) {
	if o.MaxReceiveSize != nil {
		g.MaxReceiveSize = o.MaxReceiveSize
	}
}

// UnmarshalYAML parses the images YAML
func (g *Grpc) UnmarshalYAML(node *yaml.Node) (err error) {
	if node.Kind != yaml.MappingNode {
		return errors.New(withLoc("grpc must be an object", node))
	}

	ms := node.Content
	top := len(ms)
	for i := 0; i < top; i += 2 {
		kv, err := stringKey(ms[i])
		if err != nil {
			return err
		}
		v := ms[i+1]
		switch kv {
		case "maxReceiveSize":
			val, err := resource.ParseQuantity(v.Value)
			if err != nil {
				dlog.Warningf(parseContext, "unable to parse quantity %q: %v", v.Value, withLoc(err.Error(), ms[i]))
			} else {
				g.MaxReceiveSize = &val
			}
		default:
			if parseContext != nil {
				dlog.Warn(parseContext, withLoc(fmt.Sprintf("unknown key %q", kv), ms[i]))
			}
		}
	}
	return nil
}

var parseContext context.Context

type parsedFile struct{}

func withLoc(s string, n *yaml.Node) string {
	if parseContext != nil {
		if fileName, ok := parseContext.Value(parsedFile{}).(string); ok {
			return fmt.Sprintf("file %s, line %d: %s", fileName, n.Line, s)
		}
	}
	return fmt.Sprintf("line %d: %s", n.Line, s)
}

type configKey struct{}

// WithConfig returns a context with the given Config
func WithConfig(ctx context.Context, config *Config) context.Context {
	return context.WithValue(ctx, configKey{}, config)
}

func GetConfig(ctx context.Context) *Config {
	config, ok := ctx.Value(configKey{}).(*Config)
	if !ok {
		return nil
	}
	return config
}

// GetConfigFile gets the path to the configFile as stored in filelocation.AppUserConfigDir
func GetConfigFile(c context.Context) string {
	dir, _ := filelocation.AppUserConfigDir(c)
	return filepath.Join(dir, configFile)
}

// GetDefaultConfig returns the default configuration settings
func GetDefaultConfig(c context.Context) Config {
	cfg := Config{
		Timeouts: Timeouts{
			PrivateAgentInstall:          120 * time.Second,
			PrivateApply:                 1 * time.Minute,
			PrivateClusterConnect:        20 * time.Second,
			PrivateIntercept:             5 * time.Second,
			PrivateProxyDial:             5 * time.Second,
			PrivateTrafficManagerAPI:     15 * time.Second,
			PrivateTrafficManagerConnect: 60 * time.Second,
			PrivateHelm:                  30 * time.Second,
		},
		LogLevels: LogLevels{
			UserDaemon: logrus.InfoLevel,
			RootDaemon: logrus.InfoLevel,
		},
		Cloud: Cloud{
			SkipLogin:       false,
			RefreshMessages: 24 * 7 * time.Hour,
			SystemaHost:     "app.getambassador.io",
			SystemaPort:     "443",
		},
		Grpc: Grpc{},
	}
	env := GetEnv(c)
	cfg.Images.Registry = env.Registry
	cfg.Images.WebhookRegistry = env.Registry
	cfg.Images.AgentImage = env.AgentImage
	cfg.Images.WebhookAgentImage = env.AgentImage
	return cfg
}

// LoadConfig loads and returns the Telepresence configuration as stored in filelocation.AppUserConfigDir
// or filelocation.AppSystemConfigDirs
func LoadConfig(c context.Context) (cfg *Config, err error) {
	defer func() {
		if err != nil {
			err = errcat.Config.New(err)
		}
	}()

	var dirs []string
	dirs, err = filelocation.AppSystemConfigDirs(c)
	if err != nil {
		return nil, err
	}

	dflt := GetDefaultConfig(c)
	cfg = &dflt
	readMerge := func(dir string) error {
		if stat, err := os.Stat(dir); err != nil || !stat.IsDir() { // skip unless directory
			return nil
		}
		fileName := filepath.Join(dir, configFile)
		bs, err := os.ReadFile(fileName)
		if err != nil {
			if os.IsNotExist(err) {
				err = nil
			}
			return err
		}
		parseContext = context.WithValue(c, parsedFile{}, fileName)
		defer func() {
			parseContext = nil
		}()
		fileConfig := Config{}
		if err = yaml.Unmarshal(bs, &fileConfig); err != nil {
			return err
		}
		cfg.merge(&fileConfig)
		return nil
	}

	for _, dir := range dirs {
		if err = readMerge(dir); err != nil {
			return nil, err
		}
	}
	appDir, err := filelocation.AppUserConfigDir(c)
	if err != nil {
		if os.IsNotExist(err) {
			return cfg, nil
		}
		return nil, err
	}
	if err = readMerge(appDir); err != nil {
		return nil, err
	}

	// Sanity check
	if os.Getenv("SYSTEMA_ENV") == "staging" && cfg.Cloud.SystemaHost != "beta-app.datawire.io" {
		return nil, errors.New("cloud.SystemaHost must be set to beta-app.datawire.io when using SYSTEMA_ENV set to 'staging'")
	}

	return cfg, nil
}
